### PR TITLE
Update authorization policies for SyncPlay

### PR DESCRIPTION
**Pull request by OancaAndrei**
**Changes**
Add new types of policies for accessing SyncPlay.
Add sessions counter of a user to let them access groups they've joined.

**Issues**
Related to [this](https://github.com/jellyfin/jellyfin/pull/3194#discussion_r536537652).

### CODE REVIEW

File: Jellyfin.Api/Auth/SyncPlayAccessPolicy/SyncPlayAccessHandler.cs [hunk 1]
```diff
@@ -42,10 +47,52 @@ protected override Task HandleRequirementAsync(AuthorizationHandlerContext conte
...
             var user = _userManager.GetUserById(userId!.Value);
 
-            if ((requirement.RequiredAccess.HasValue && user.SyncPlayAccess == requirement.RequiredAccess)
-                || user.SyncPlayAccess == SyncPlayAccess.CreateAndJoinGroups)
+            if (requirement.RequiredAccess == SyncPlayAccessRequirementType.HasAccess)
             {
-                context.Succeed(requirement);
+                if (user.SyncPlayAccess == SyncPlayUserAccessType.CreateAndJoinGroups ||
+                    user.SyncPlayAccess == SyncPlayUserAccessType.JoinGroups ||
+                    _syncPlayManager.IsUserActive(userId!.Value))
```
Comment (crobibero) [inline Jellyfin.Api/Auth/SyncPlayAccessPolicy/SyncPlayAccessHandler.cs:54]:
```suggestion
                if (user.SyncPlayAccess == SyncPlayUserAccessType.CreateAndJoinGroups
                    || user.SyncPlayAccess == SyncPlayUserAccessType.JoinGroups
                    || _syncPlayManager.IsUserActive(userId!.Value))
```

File: Jellyfin.Api/Auth/SyncPlayAccessPolicy/SyncPlayAccessHandler.cs [hunk 2]
```diff
@@ -42,10 +47,52 @@ protected override Task HandleRequirementAsync(AuthorizationHandlerContext conte
...
+                }
+                else
+                {
+                    context.Fail();
+                }
+            }
+            else if (requirement.RequiredAccess == SyncPlayAccessRequirementType.JoinGroup)
+            {
+                if (user.SyncPlayAccess == SyncPlayUserAccessType.CreateAndJoinGroups ||
+                    user.SyncPlayAccess == SyncPlayUserAccessType.JoinGroups)
```
Comment (crobibero) [inline Jellyfin.Api/Auth/SyncPlayAccessPolicy/SyncPlayAccessHandler.cs:77]:
```suggestion
                if (user.SyncPlayAccess == SyncPlayUserAccessType.CreateAndJoinGroups
                    || user.SyncPlayAccess == SyncPlayUserAccessType.JoinGroups)
```

**Review by crobibero — APPROVED** [review]
Should this be ported to 10.7?

**Comment by OancaAndrei** [comment]
> Should this be ported to 10.7?

Yes, please. :+1:

**Review by BaronGreenback — APPROVED** [review]

File: Jellyfin.Api/Auth/SyncPlayAccessPolicy/SyncPlayAccessHandler.cs [hunk 3]
```diff
@@ -42,10 +47,52 @@ protected override Task HandleRequirementAsync(AuthorizationHandlerContext conte
...
             var user = _userManager.GetUserById(userId!.Value);
 
-            if ((requirement.RequiredAccess.HasValue && user.SyncPlayAccess == requirement.RequiredAccess)
-                || user.SyncPlayAccess == SyncPlayAccess.CreateAndJoinGroups)
+            if (requirement.RequiredAccess == SyncPlayAccessRequirementType.HasAccess)
             {
-                context.Succeed(requirement);
+                if (user.SyncPlayAccess == SyncPlayUserAccessType.CreateAndJoinGroups
+                    || user.SyncPlayAccess == SyncPlayUserAccessType.JoinGroups
+                    || _syncPlayManager.IsUserActive(userId!.Value))
```
Comment (cvium) [inline Jellyfin.Api/Auth/SyncPlayAccessPolicy/SyncPlayAccessHandler.cs:54]: I think a better, albeit more involved and fragile solution is to have the syncplay session id in the route data and use it to verify that the active user is part of _that_ group.

Alternatively, do the check explicitly in the controller eg.
```
if (!_syncPlayManager.IsUserInGroup(userId, groupId))
{
    return Forbid(); // or BadRequest("User cannot alter the state of another group");
}
```

File: Jellyfin.Api/Auth/SyncPlayAccessPolicy/SyncPlayAccessHandler.cs [hunk 4]
```diff
@@ -42,10 +47,52 @@ protected override Task HandleRequirementAsync(AuthorizationHandlerContext conte
             var userId = ClaimHelpers.GetUserId(context.User);
             var user = _userManager.GetUserById(userId!.Value);
 
-            if ((requirement.RequiredAccess.HasValue && user.SyncPlayAccess == requirement.RequiredAccess)
-                || user.SyncPlayAccess == SyncPlayAccess.CreateAndJoinGroups)
+            if (requirement.RequiredAccess == SyncPlayAccessRequirementType.HasAccess)
```
Comment (cvium) [inline Jellyfin.Api/Auth/SyncPlayAccessPolicy/SyncPlayAccessHandler.cs:50]: I don't know if it's too late to change it, but if the SyncPlayUserAccessType has the `[Flag]` attribute, then you could use bit masking and avoid the `SyncPlayAccessRequirementType` enum.

**Comment by joshuaboniface** [comment]
@OancaAndrei re: @cvium's feedback, that might warrant another PR but I'll leave you guys to discuss it on Matrix.

File: Jellyfin.Api/Auth/SyncPlayAccessPolicy/SyncPlayAccessHandler.cs (hunk 4 above)
Comment (OancaAndrei) [reply Jellyfin.Api/Auth/SyncPlayAccessPolicy/SyncPlayAccessHandler.cs:50]: Bit masking does indeed sound more appropriate for this, I'll have to look into it.

**Pull request merged**